File: RotateArray.java

import java.util.Arrays;

public class RotateArray {

    // Helper function to reverse elements of the array
    public static void reverse(int[] arr, int start, int end) {
        while (start < end) {
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }
    }

    // Function to rotate the array by k positions
    public static void rotate(int[] arr, int k) {
        int n = arr.length;
        k = k % n;  // In case k is greater than array length

        // Step 1: Reverse the whole array
        reverse(arr, 0, n - 1);

        // Step 2: Reverse the first k elements
        reverse(arr, 0, k - 1);

        // Step 3: Reverse the remaining elements
        reverse(arr, k, n - 1);
    }

    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6, 7, 8};
        int k = 4;

        System.out.println("Original Array: " + Arrays.toString(arr));

        rotate(arr, k);

        System.out.println("Array after rotating by " + k + " positions: " + Arrays.toString(arr));
    }
}

File: javascript/rotate-array.js
/*
    rotate an array by k positions
 */


const input = [1, 2, 3, 4, 5, 6, 7]
// result = [4,5,6,7,8,1,2,3]
const k = 3;
function reverseArray(input, start, end) {
    while (start < end) {
        temp = input[start]
        /*  s=0, e=7
            s=1,e=6
            s=2,e=5,
            s=3,e=4
            s=4,e=3
        */
        input[start] = input[end]
        input[end] = temp
        start++
        end--
    }

}


function rotateArray(input, k) {

    const start = 0;
    const end = input.length - 1;

    reverseArray(input, start, end)
    console.log(`first reverse` + input)
    reverseArray(input, start, k - 1)
    console.log(`second reverse` + input)
    reverseArray(input, k, end)
    console.log(`ROTATED ARRAY` + input)

}
rotateArray(input, k)